pygdf/_gdf.py

```diff
@@ -239,6 +256,92 @@ def apply_join(col_lhs, col_rhs, how, method='hash'):
     libgdf.gdf_column_free(col_result_r)
 
 
+def libgdf_join(col_lhs, col_rhs, on, how, method='hash'):
+
+    joiner = _join_how_api[how]
+    method_api = _join_method_api[method]
+    gdf_context = ffi.new('gdf_context*')
+
+    if method == 'hash':
+        libgdf.gdf_context_view(gdf_context, 0, method_api, 0)
+    elif method == 'sort':
+        libgdf.gdf_context_view(gdf_context, 1, method_api, 0)
+    else:
+        msg = "method not supported"
+        raise ValueError(msg)
+
+    if(how not in ['left', 'inner']):
+        msg = "new join api only supports left or inner"
+        raise ValueError(msg)
+
+    list_lhs = []
+    list_rhs = []
+    result_cols = []
+
+    result_col_names = []
+
+    left_idx = []
+    idx = 0
+    for name, col in col_lhs.items():
+
+        list_lhs.append(col._column.cffi_view)
+        if name not in on:
+            result_cols.append(columnview(0, None, dtype=col._column.dtype))
+            result_col_names.append(name)
+        else:
+            left_idx.append(idx)
+        idx = idx + 1
+
+    for name in on:
+        result_cols.append(columnview(0, None,
+                                      dtype=col_lhs[name]._column.dtype))
+        result_col_names.append(name)
+
+    # right_idx = ffi.new('int[' + str(len(on)) + ']')
+    right_idx = []
+    idx = 0
+    for name, col in col_rhs.items():
+        list_rhs.append(col._column.cffi_view)
+
+        if name not in on:
+            result_cols.append(columnview(0, None, dtype=col._column.dtype))
+            result_col_names.append(name)
+        else:
+            right_idx.append(idx)
```

Comment:
Does the libgdf call expect the indices to be the same for the matching columns? I.E. left_idx[0] and right_idx[0] correspond to the matching column.

If so, if the columns aren't in the same order this won't match up properly.